lib/streamlit/runtime.py

```diff
@@ -0,0 +1,591 @@
+        )
+
+    @property
+    def state(self) -> RuntimeState:
+        return self._state
+
+    @property
+    def message_cache(self) -> ForwardMsgCache:
+        return self._message_cache
+
+    @property
+    def uploaded_file_mgr(self) -> UploadedFileManager:
+        return self._uploaded_file_mgr
+
+    @property
+    def stats_mgr(self) -> StatsManager:
+        return self._stats_mgr
+
+    def _on_files_updated(self, session_id: str) -> None:
+        """Event handler for UploadedFileManager.on_file_added.
+        Ensures that uploaded files from stale sessions get deleted.
+
+        Threading
+        ---------
+        SAFE. May be called on any thread.
+        """
+        session_info = self._get_session_info(session_id)
+        if session_info is None:
+            # If an uploaded file doesn't belong to an existing session,
+            # remove it so it doesn't stick around forever.
+            self._uploaded_file_mgr.remove_session_files(session_id)
+
+    def _get_session_info(self, session_id: str) -> Optional[SessionInfo]:
+        """Return the SessionInfo with the given id, or None if no such
+        session exists.
+
+        Threading
+        ---------
+        UNSAFE. Must be called on the eventloop thread.
+        """
+        return self._session_info_by_id.get(session_id, None)
+
+    async def start(self, on_started: Optional[Callable[[], Any]] = None) -> None:
+        """Start the runtime. This must be called only once, before
+        any other functions are called.
+
+        Parameters
+        ----------
+        on_started
+            An optional callback that will be called when the runtime's loop
+            has started. It will be called on the eventloop thread.
+
+        Returns
+        -------
+        None
+
+        Threading
+        ---------
+        UNSAFE. Must be called on the eventloop thread.
+        """
+        await self._loop_coroutine(on_started)
+
+    def stop(self) -> None:
+        """Request that Streamlit close all sessions and stop running.
+        Note that Streamlit won't stop running immediately.
+
+        Threading
+        ---------
+        SAFE. May be called on any thread.
+        """
+        if self._state in (RuntimeState.STOPPING, RuntimeState.STOPPED):
+            return
+
+        LOGGER.debug("Runtime stopping...")
+        self._set_state(RuntimeState.STOPPING)
+        self._get_eventloop().call_soon_threadsafe(self._must_stop.set)
+
+    def is_active_session(self, session_id: str) -> bool:
+        """True if the session_id belongs to an active session.
+
+        Threading
+        ---------
+        SAFE. May be called on any thread.
+        """
+        # Dictionary membership is atomic in CPython, so this is thread-safe.
+        return session_id in self._session_info_by_id
+
+    def create_session(
+        self,
+        client: SessionClient,
+        user_info: Dict[str, Optional[str]],
+    ) -> str:
```

Comment:
Session creation and management now operates on session ID strings rather than `SessionInfo` instances. `SessionInfo` is an opaque implementation detail of `Runtime`, and we don't expose `SessionInfo` or `AppSession` to the outside world. (Only the Runtime object calls functions on `AppSession`.)